pymc_experimental/distributions/timeseries.py

```diff
@@ -0,0 +1,198 @@
+    Checks that P is a valid transition matrix
+    """
+
+    # TODO: Can this eval be avoided?
+    n, k = P.shape.eval()
+    if n != k:
+        raise ValueError(f'P must be square, found shape ({n}, {k})')
+
+    row_sums_all_one = at.allclose(P.sum(axis=1), 1.0)
+    if not row_sums_all_one.eval():
+        raise ValueError('All rows of P must sum to 1.')
+
+
+class DiscreteMarkovChainRV(SymbolicRandomVariable):
+    default_output = 1
+    _print_name = ('DiscreteMC', '\\operatorname{DiscreteMC}')
+
+    def update(self, node: Node):
+        # TODO: Do I need this?
+        return {node.inputs[-1]: node.outputs[0]}
+
+
+class DiscreteMarkovChain(Distribution):
+    r"""
+    A Discrete Markov Chain is a sequence of random variables
+    .. math::
+        \{x_t\}_{t=0}^T
+    Where transition probability P(x_t | x_{t-1}) depends only on the state of the system at x_{t-1}.
+
+    Parameters
+    ----------
+    P: tensor
+        Matrix of transition probabilities between states. Rows must sum to 1.
+        One of P or P_logits must be provided.
+    P_logit: tensor, Optional
+        Matrix of tranisiton logits. Converted to probabilities via Softmax activation.
+        One of P or P_logits must be provided.
+    steps: tensor
+        Length of the markov chain
+    x0: tensor or RandomVariable
+        Intial state of the system. If tensor, treated as deterministic.
+    """
+
+    rv_type = DiscreteMarkovChainRV
+
+    def __new__(cls, *args, steps, **kwargs):
+        # TODO: Allow steps to be None and infer chain length from shape?
+        # TODO: Dims breaks the RV
+
+        # Subtract 1 step to account for x0 given, better match user expectation of
+        # len(markov_chain) = steps
+        steps -= 1
+
+        steps = get_support_shape_1d(
+            support_shape=steps,
+            shape=None,
+            dims=kwargs.get('dims', None),
+            observed=kwargs.get('observed', None),
+            support_shape_offset=1
+        )
+
+        return super().__new__(cls, *args, steps=steps, **kwargs)
+
+    @classmethod
+    def dist(cls, P=None, logit_P=None, steps=None, x0=None, **kwargs):
+        steps = get_support_shape_1d(
+            support_shape=steps, shape=kwargs.get('shape', None), support_shape_offset=1
+        )
+        if steps is None:
+            raise ValueError("Must specify steps or shape parameter")
+        if P is None and logit_P is None:
+            raise ValueError('Must specify P or logit_P parameter')
+        if P is not None and logit_P is not None:
+            raise ValueError('Must specify only one of either P or logit_P parameter')
+
+        if logit_P is not None:
+            P = pm.math.softmax(logit_P, axis=1)
+        P = at.as_tensor_variable(P)
+        validate_transition_matrix(P)
+
+        # TODO: Can this eval be avoided?
+        n_states = P.shape[0].eval()
+
+        if not isinstance(x0, TensorVariable):
+            x0 = at.as_tensor_variable(x0).astype(intX)
```

Comment:
If x0 is accounted for in the logp, you should treat it like init_dist in other timeseries: check it's not registered in the model, resize it to match the size (batched dims) and call `ignore_logprob` so PyMC doesn't complain when it sees a "loose" RV without a value.

Reply:
x0 isn't included in the logp since you can't get the probability of x0 out of the transition matrix (no associated previous value). The first term of the logp is `log(P[state_0, state_1])`, which is P(state_1 | state_0).

I had hoped with this setup everything would magically work. It doesn't complain about loose RVs, and the x0 probability should be automatically handled by the prior on x0. If you check the notebook I did have an issue where the individual logp terms didn't add up as expected, though -- not sure if this is related.

Reply:
Okay. You can consider including it, so that users don't have to worry about shaping issues. Then they would call `DiscreteMarkov(init_dist=Categorical.dist(), P=..., shape=...)` And because you have control over the size of init_dist, some of the checks can probably go away.

Reply:
On the other hand it makes it difficult to have constant x0, if that's desired.

Reply:
I split them on the basis of "clean coding": since the x0 distribution isn't, strictly speaking, part of the markov chain, there's no reason for the markov chain class to be creating it. 

What's more important though is that this matches the API of other distributions that ask for an initial condition. I would change this to match the way you and the other core devs consider too be the most "PyMC-like".

Reply:
The integrated init_dist is more in line with our timeseries. I don't particularly love the API, but I guess it's better to stick with it until someone thinks of a comprehensive alternative.

Reply:
I'll make the change. IMO the comprehensive alternative is to explicitly demand the user make their own initial distributions, since all of these distributions are written mathematically with "x0 given" .

Reply:
x0 parameter removed, added `init_dist` parameter that takes an unnamed distribution as an argument. Docstring and type checks were copied from `pymc.distributions.timeseries.AR`

Reply:
By comprehensive I mostly meant user friendly :)

Reply:
I guess to me the alternative is that the init_dist parameters want actual random variables, not `.dist`. They can still be concatenated to the time series output internally -- your point Pro (1) is very crucial I think and any implementation should have this property. I would argue that Con (1) applies more to my proposal, because you have to do a bunch of type and shape checking on the distribution being provided. But these checks could be farmed out to a function common to all time series distributions, so the individual distributions wouldn't need to know anything about it.

To me the biggest advantage is allowing the graph to naturally handle the extra x0 logp, rather than having to manually ignore it in the `.dist` method and then re-compute it in the `.logp` method.